plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -172,7 +172,7 @@ export class KubernetesProxy {
             )?.toString(),
           };
 
-          const authHeader = req.header(HEADER_KUBERNETES_AUTH);
+          const authHeader = req.header?.(HEADER_KUBERNETES_AUTH);
```

Comment:
I'm still not sure I understand WHY this would ever be undefined. Feels like we're missing something here.

Also could this just have been `req.headers[HEADER_KUBERNETES_AUTH]` instead?

Reply:
When I run the backstage app with the Kubernetes plugin installed, sometimes I got the error, and it is quite frequently.
req.header is a function, and the header might be null, that is why I has this pr.

Reply:
Yeah the function might be entirely missing it seems. Which sounds really really bad. I'm just worried we might be glossing over some much worse underlying problem, which is why I'm pondering this a bit extra

Reply:
Could you change it to instead call `req.get(HEADER_KUBERNETES_AUTH)`? 

https://expressjs.com/en/api.html

Reply:
Why? I believe it will be the same: get is the same as header function, and get might be null as well

Reply:
Maybe, maybe not.

Do you have ANY idea why this might happen?

(note that the changeset does need to be fixed too)

Reply:
I believe this can actually be a plain `IncomingMessage` from `'http'` under the hood, and that the types in the upstream package are incorrect. The safest fix here would be to refactor this code to use `req.headers` instead, since that's the LCD

Reply:
Alright so,

```suggestion
          const authHeader = req.headers[HEADER_KUBERNETES_AUTH];
```

Reply:
> Maybe, maybe not.
> 
> Do you have ANY idea why this might happen?
> 
> (note that the changeset does need to be fixed too)

@freben, req.header is alias for req.get, https://expressjs.com/en/api.html#req.get,  I will fix the changeset as you suggested, thanks

Reply:
> I believe this can actually be a plain `IncomingMessage` from `'http'` under the hood, and that the types in the upstream package are incorrect. The safest fix here would be to refactor this code to use `req.headers` instead, since that's the LCD

@Rugvip, @freben  I am afraid this might now work, what if headers is undefined? do we have to use `req.headers?.[HEADER_KUBERNETES_AUTH]` instead?

Reply:
@Rugvip I just confirmed that after I changed to req.headers, then the 2 tests are failing, and I got compilation error warning in my vs code IDE:
<img width="801" alt="Screenshot 2024-04-09 at 10 35 41 AM" src="https://github.com/backstage/backstage/assets/152627/7af1b167-eca5-4bdd-b2a7-263f75140d4b">
and the type script error is like this:
```
Type 'string | string[]' is not assignable to type 'string | undefined'.
  Type 'string[]' is not assignable to type 'string'.ts(2322)
```
Basically in line 175, the authHeader  type is string | string [], while req.headers.authorization type is string | undefined, so they are not compatible.

So should I keep using the req.header?.() anyway?

Reply:
That sounds more complicated, could you explain a little bit why you prefer req.headers[] over req.header()?

Reply:
> Yeah because it supports multiple headers upon reading, so you'll have to deal with that.
> 
> 
> 
> Just simply `[authHeader].flat()[0]` will do, I'd say (i.e. take the first one if present), or be thorough and make an if statement that checks that it has exactly one or zero entries etc.

Then how to handle the case that req.headers is null?

Reply:
The theory is that `req.headers` WILL exist (it's part of the low level type contract) while `req.header` won't (it's part of the express layer on top, and it seems sometimes that's not the type actually received).

Reply:
Actually I tried to use req.headers yesterday and got many tests failed, later I figured out that the string value has to be lowercase, https://stackoverflow.com/questions/60855411/req-header-vs-req-headers-in-express